python/lsst/ip/diffim/dipoleFitTask.py

```diff
@@ -0,0 +1,921 @@
+
+        return zout
+
+    def fitDipoleImpl(self, source, tol=1e-7, rel_weight=0.5,
+                      fitBgGradient=True, bgGradientOrder=1, maxSepInSigma=5.,
+                      separateNegParams=True, verbose=False, display=False):
+        """Fit a dipole model to an input difference image (actually,
+        subimage bounded by the input source's footprint) and
+        optionally constrain the fit using the pre-subtraction images
+        posImage and negImage.
+
+        @see `fitDipole()`
+        @return `lmfit.MinimizerResult` object containing the fit
+        parameters and other information.
+        """
+
+        # Only import lmfit if someone wants to use the new DipoleFitAlgorithm.
+        import lmfit
+
+        fp = source.getFootprint()
+        bbox = fp.getBBox()
+        subim = afwImage.MaskedImageF(self.diffim.getMaskedImage(), bbox, afwImage.PARENT)
+
+        z = diArr = subim.getArrays()[0]
+        weights = 1. / subim.getArrays()[2]  # get the weights (=1/variance)
+        if self.posImage is not None and rel_weight > 0.:
+            posSubim = afwImage.MaskedImageF(self.posImage.getMaskedImage(), bbox, afwImage.PARENT)
+            if self.negImage is not None:
+                negSubim = afwImage.MaskedImageF(self.negImage.getMaskedImage(), bbox, afwImage.PARENT)
+            else:  # no template provided; generate it from the obs. and diffim
+                negSubim = posSubim.clone()
+                negSubim -= subim
+            z = np.append([z], [posSubim.getArrays()[0],
+                                negSubim.getArrays()[0]], axis=0)
+            # Weight the pos/neg images by rel_weight relative to the diffim
+            weights = np.append([weights], [1. / posSubim.getArrays()[2] * rel_weight,
+                                            1. / negSubim.getArrays()[2] * rel_weight], axis=0)
+        else:
+            rel_weight = 0.  # a short-cut for "don't include the pre-subtraction data"
+
+        # Create the lmfit model (lmfit uses scipy 'leastsq' option by default - Levenberg-Marquardt)
+        gmod = lmfit.Model(DipoleFitAlgorithm.genDipoleModel, verbose=verbose)
+
+        # Add the constraints for centroids, fluxes.
+        # starting constraint - near centroid of footprint
+        fpCentroid = np.array([fp.getCentroid().getX(), fp.getCentroid().getY()])
+        cenNeg = cenPos = fpCentroid
+
+        pks = fp.getPeaks()
+        if len(pks) >= 1:
+            cenPos = pks[0].getF()    # if individual (merged) peaks were detected, use those
+        if len(pks) >= 2:
+            cenNeg = pks[1].getF()
+
+        # For close/faint dipoles the starting locs (min/max) might be way off, let's help them a bit.
+        # First assume dipole is not separated by more than 5*psfSigma.
+        maxSep = self.psfSigma * maxSepInSigma
+
+        # Note - this may be a cheat to assume the dipole is centered in center of the footprint.
```

Comment:
It's a reasonable starting point, but it is not an overall constraint, so I modified the comment.